File: python/core/Node.py

from typing import Optional, List, Dict, Any, Type, Callable, TYPE_CHECKING
import sys
import logging
from enum import Enum, auto

from abc import ABC, abstractmethod

import uuid
import random

# To avoid circular imports
if TYPE_CHECKING:
    from NodePort import NodePort
  
from .Interface import IExecutionContext, IExecutionResult, INode, INodePort, IInputControlPort, IOutputControlPort, IInputDataPort, IOutputDataPort

from .Types import ValueType, PortFunction, NodeKind
from .NodePort import NodePort, InputDataPort, OutputDataPort, InputControlPort, OutputControlPort
from .GraphPrimitives import GraphNode
# Get a logger for this module
logger = logging.getLogger(__name__)


class PluginRegistry:
    _registry: Dict[str, Type['Node']] = {}

    @classmethod
    def register(cls, type_name: str) -> Callable[[Type['Node']], Type['Node']]:
        """Decorator to register a node class with a specific type name."""
        def decorator(subclass: Type['Node']) -> Type['Node']:
            if cls._registry.get(type_name):
                raise ValueError(f"Node type '{type_name}' is already registered.")
            cls._registry[type_name] = subclass
            return subclass
        return decorator
    
    @classmethod
    def get_node_class(cls, type_name: str) -> Optional[Type['Node']]:
        return cls._registry.get(type_name)
    
    @classmethod
    def create_node(cls, node_id: str, type_name: str, *args, **kwargs) -> 'Node':
        """Factory method to create a node instance by type name."""
        NodeClass = cls.get_node_class(type_name)
        if not NodeClass:
            raise ValueError(f"Unknown node type '{type_name}'")
        return NodeClass(node_id, type_name, *args, **kwargs)
    
    def get_registered_types(cls) -> List[str]:
        return list(cls._registry.keys())
    


# A typescript map behaves like an ordered dict in python
class Node(INode):
    _node_registry: Dict[str, Type['Node']] = {}

    @classmethod
    def register(cls, type_name: str) -> Callable[[Type['Node']], Type['Node']]:
        """Decorator to register a node class with a specific type name."""
        def decorator(subclass: Type['Node']) -> Type['Node']:
            if cls._node_registry.get(type_name):
                raise ValueError(f"Node type '{type_name}' is already registered.")
            cls._node_registry[type_name] = subclass
            return subclass
        return decorator

    @classmethod
    def create_node(cls, node_id: str, type_name: str, *args, **kwargs) -> 'Node':
        """Factory method to create a node instance by type name."""
        # Use simple dictionary lookup for O(1) factory
        if type_name not in cls._node_registry:
            # Fallback or strict error? 
            # For now, let's just try to be helpful or error out.
            raise ValueError(f"Unknown node type '{type_name}'")
        
        node_class = cls._node_registry[type_name]
        return node_class(node_id, type_name, *args, **kwargs)

    def __init__(self, 
                 name: str, 
                 type: str, 
                 network_id: str,
                 inputs: Optional[Dict[str, NodePort]] = None, 
                 outputs: Optional[Dict[str, NodePort]] = None):
        
        #super().__init__(name, type, network.id if network else None)
        self.name = name
        self.id = uuid.uuid4().hex
        #self.id = uuid.uuid4().hex  # Unique identifier for the node
        self.type = type
        self.uuid = uuid.uuid4().hex  # Unique identifier for the node instance

        # Use Standard Dicts - Explicit Dict[str, NodePort] for TS Map<string, NodePort>
        self.inputs: Dict[str, NodePort] = inputs if inputs is not None else {}
        self.outputs: Dict[str, NodePort] = outputs if outputs is not None else {}

        self.is_flow_control_node = False
        self.is_loop_node = False 
        self._isDirty = True  
        self.network_id = network_id
        self.graph = None # Will be set by the NodeNetwork when added
        self.path= " path node computed at runtime " # This is a bit of a hack, but it allows us to have a path property that is computed at runtime based on the network structure. In Rust/TS, we can compute this on demand or cache it as needed.

        self.kind = NodeKind.FUNCTION

   

    def isNetwork(self) -> bool:
        return self.kind == NodeKind.NETWORK

    def isDataNode(self) -> bool:
        return self.is_flow_control_node == False

    def isFlowControlNode(self) -> bool:
        return self.is_flow_control_node == True
    
    def markDirty(self):
        self._isDirty = True
        
    def markClean(self):
        self._isDirty = False

    def isDirty(self) -> bool:  
        return self._isDirty

    # TODO: NOT AT ALL TESTED
    def delete_input(self, port_name: str):
        if port_name in self.inputs:
            # TODO: logic omitted for brevity in porting step
            # Note: Removal needs to handle connection cleanup explicitly in Rust/TS
            port = self.inputs[port_name]
            # ... cleanup code ...
            del self.inputs[port_name]    

    def delete_output(self, port_name: str):
        if port_name in self.outputs:
            del self.outputs[port_name]

    
    def add_control_input(self, port_name: str) -> InputControlPort:
        port = InputControlPort(self.id, port_name)
        self.inputs[port_name] = port
        return port
    

    def add_control_output(self, port_name: str) -> OutputControlPort:
        port = OutputControlPort(self.id, port_name)
        self.outputs[port_name] = port
        return port
    

    def add_data_input(self, port_name: str, data_type: ValueType = ValueType.ANY) -> InputDataPort:
        if port_name in self.inputs:
            raise ValueError(f"Data input port '{port_name}' already exists in node '{self.id}'")
        
        port = InputDataPort(self.id, port_name, data_type=data_type)
        self.inputs[port_name] = port

        return port

    
    def add_data_output(self, port_name: str, data_type: ValueType = ValueType.ANY) -> OutputDataPort:
        if port_name in self.outputs:
            raise ValueError(f"Data output port '{port_name}' already exists in node '{self.id}'")
        
        port = OutputDataPort(self.id, port_name, data_type=data_type)
        self.outputs[port_name] = port

        return port


    # Updated helper to use Enum instead of Int Flag
    def get_input_ports(self, restrict_to: Optional[PortFunction] = None) -> List[NodePort]:
        if restrict_to is None:
            return list(self.inputs.values())
            
        return [port for port in self.inputs.values() if port.function == restrict_to]

            
    def get_output_ports(self, restrict_to: Optional[PortFunction] = None) -> List[NodePort]:
        if restrict_to is None:
            return list(self.outputs.values())
            
        return [port for port in self.outputs.values() if port.function == restrict_to]
            
    def get_output_data_ports(self) -> List[NodePort]:
        return self.get_output_ports(restrict_to=PortFunction.DATA)

    def get_output_control_ports(self) -> List[NodePort]:
        return self.get_output_ports(restrict_to=PortFunction.CONTROL)  

    def get_input_data_ports(self) -> List[NodePort]:
        return self.get_input_ports(restrict_to=PortFunction.DATA)

   
    def get_input_control_ports(self) -> List[NodePort]:
        return self.get_input_ports(restrict_to=PortFunction.CONTROL)
    
        

    def get_input_data_port(self, port_name: str) -> NodePort: 
        port = self.inputs.get(port_name)
        
        if not port:
            raise KeyError(f"Input port '{port_name}' not found in node '{self.id}'")
        if not port.isDataPort():
            raise KeyError(f"Input port '{port_name}' in node '{self.id}' is not a data port")
        
        return port
    
    def get_output_data_port(self, port_name: str) -> NodePort:
        port = self.outputs.get(port_name)
        if hasattr(port, 'isDataPort'): assert(port.isDataPort()) # type: ignore
        if not port:
            raise KeyError(f"Output port '{port_name}' not found in node '{self.id}'")
        if not port.isDataPort():
            raise KeyError(f"Output port '{port_name}' in node '{self.id}' is not a data port")
        
        return port
    
    def get_input_control_port(self, port_name: str) -> NodePort:
        port = self.inputs.get(port_name)
        if hasattr(port, 'isControlPort'): assert(port.isControlPort()) # type: ignore
        if not port:
            raise ValueError(f"Input port '{port_name}' not found in node '{self.id}'")
        if not port.isControlPort():
            raise ValueError(f"Input port '{port_name}' in node '{self.id}' is not a control port")
        
        return port
    
    def get_output_control_port(self, port_name: str) -> NodePort:
        port = self.outputs.get(port_name)
        if hasattr(port, 'isControlPort'): assert(port.isControlPort()) # type: ignore
        if not port:
            raise ValueError(f"Output port '{port_name}' not found in node '{self.id}'")
        if not port.isControlPort():
            raise ValueError(f"Output port '{port_name}' in node '{self.id}' is not a control port")
        
        return port


    # TODO: precompute should not compute inputs. this should be done in compute() and this should be a callback only
    def precompute(self):
       print(f"PRECOMPUTE: node '{self.id}' ")


    def postcompute(self):
        logger.info(f"POSTCOMPUTE: node '{self.id}' ")
        # validation checks. We're really just making sure the node is working correclty here.
        self.dump_dirty_states()
        assert(self.all_data_inputs_clean())
        
        assert(self.all_data_outputs_clean())


        self._isDirty = False
        


    # this is checking for cleen DATA inputs only
    def all_data_inputs_clean(self):
        for input_port in self.get_input_data_ports():
            if input_port._isDirty:
                return False
        return True
    

    def all_data_outputs_clean(self):
        for output_port in self.get_output_data_ports():
            if output_port._isDirty:
                return False
        return True
    
    def dump_dirty_states(self):
    
        print(f"  Node '{self.id}' port dirty states:")
        print(f".    Control Inputs: ({len(self.get_input_control_ports())})")
        for input_port in self.get_input_control_ports():
            print(f".       Input Port [{input_port.port_name}] dirty: {input_port.isDirty()}; active: {input_port.isActive()}")
        print(f".    Control Outputs: ({len(self.get_output_control_ports())})")
        for output_port in self.get_output_control_ports():
            print(f".       Output Port [{output_port.port_name}] dirty: {output_port.isDirty()}; active: {output_port.isActive()}")
        print(f".    Data Inputs: ({len(self.get_input_data_ports())})")
        for input_port in self.get_input_data_ports():
            print(f".       Input Port [{input_port.port_name}] dirty: {input_port.isDirty()}; value: {input_port.value}")
        print(f".    Data Outputs: ({len(self.get_output_data_ports())})")
        for output_port in self.get_output_data_ports():
            print(f".       Output Port [{output_port.port_name}] dirty: {output_port.isDirty()}; value: {output_port.value}")
    

    @staticmethod
    @abstractmethod
    async def compute(executionContext) -> IExecutionResult:
        pass

    def compile(self, builder: Any):
        """
        [Compile Phase]
        Does NOT run the node. Instead, emits Intermediate Representation (IR) instructions
        to the 'builder'. This IR allows generation of AssemblyScript, WASM, or Optimized Rust.
        """
        pass
       
    
        # if any inputs are still dirty, we cannot compute this node
        # if any input ports have changed then recompute.

        pass  # Placeholder for node computation logic

    def generate_IRC(self):
        # Placeholder for generating intermediate representation code for this node
        pass

File: typescript/src/core/Node.ts
import { v4 as uuidv4 } from 'uuid';
import {
  IExecutionContext,
  IExecutionResult,
  INode,
  INodePort,
  IInputControlPort,
  IOutputControlPort,
  IInputDataPort,
  IOutputDataPort,
} from './Interface';
import { ValueType, PortFunction, NodeKind } from './Types';
import {
  NodePort,
  InputDataPort,
  OutputDataPort,
  InputControlPort,
  OutputControlPort,
} from './NodePort';
import { GraphNode } from './GraphPrimitives';

// ─────────────────────────────────────────────
// PluginRegistry — mirrors Python PluginRegistry
// ─────────────────────────────────────────────

export class PluginRegistry {
  private static _registry: Map<string, typeof Node> = new Map();

  /** Decorator factory to register a node class with a specific type name. */
  static register(typeName: string) {
    return function <T extends typeof Node>(subclass: T): T {
      if (PluginRegistry._registry.has(typeName)) {
        throw new Error(`Node type '${typeName}' is already registered.`);
      }
      PluginRegistry._registry.set(typeName, subclass as unknown as typeof Node);
      return subclass;
    };
  }

  static get_node_class(typeName: string): typeof Node | undefined {
    return PluginRegistry._registry.get(typeName);
  }

  static create_node(nodeId: string, typeName: string, ...args: any[]): Node {
    const NodeClass = PluginRegistry.get_node_class(typeName);
    if (!NodeClass) throw new Error(`Unknown node type '${typeName}'`);
    return new (NodeClass as any)(nodeId, typeName, ...args);
  }

  static get_registered_types(): string[] {
    return Array.from(PluginRegistry._registry.keys());
  }
}

// ─────────────────────────────────────────────
// Node — base class for all computation nodes
// ─────────────────────────────────────────────

export abstract class Node extends INode {
  // Separate registry from PluginRegistry (mirrors Python Node._node_registry)
  static _node_registry: Map<string, typeof Node> = new Map();

  /** Decorator factory to register a node class with a specific type name. */
  static register(typeName: string) {
    return function <T extends typeof Node>(subclass: T): T {
      if (Node._node_registry.has(typeName)) {
        throw new Error(`Node type '${typeName}' is already registered.`);
      }
      Node._node_registry.set(typeName, subclass as unknown as typeof Node);
      return subclass;
    };
  }

  static create_node(nodeId: string, typeName: string, ...args: any[]): Node {
    if (!Node._node_registry.has(typeName)) {
      throw new Error(`Unknown node type '${typeName}'`);
    }
    const NodeClass = Node._node_registry.get(typeName)!;
    return new (NodeClass as any)(nodeId, typeName, ...args);
  }

  // ─── Instance state ───
  name: string;
  id: string;
  uuid: string;
  type: string;
  network_id: string | null;
  inputs: Record<string, NodePort>;
  outputs: Record<string, NodePort>;
  is_flow_control_node: boolean;
  is_loop_node: boolean;
  _isDirty: boolean;
  graph: any; // set by NodeNetwork when added
  path: string;
  kind: NodeKind;

  constructor(
    name: string,
    type: string,
    networkId: string | null = null,
    inputs?: Record<string, NodePort>,
    outputs?: Record<string, NodePort>,
    /** absorb extra kwargs forwarded from subclasses */
    ...rest: any[]
  ) {
    super();
    this.name = name;
    this.id = uuidv4().replace(/-/g, '');
    this.uuid = uuidv4().replace(/-/g, '');
    this.type = type;
    this.network_id = networkId;
    this.inputs = inputs ?? {};
    this.outputs = outputs ?? {};
    this.is_flow_control_node = false;
    this.is_loop_node = false;
    this._isDirty = true;
    this.graph = null;
    this.path = ' path node computed at runtime ';
    this.kind = NodeKind.FUNCTION;
  }

  isNetwork(): boolean {
    return this.kind === NodeKind.NETWORK;
  }

  isDataNode(): boolean {
    return !this.is_flow_control_node;
  }

  isFlowControlNode(): boolean {
    return this.is_flow_control_node;
  }

  markDirty(): void {
    this._isDirty = true;
  }

  markClean(): void {
    this._isDirty = false;
  }

  isDirty(): boolean {
    return this._isDirty;
  }

  /**
   * Snapshot this node's internal execution state for checkpoint/resume.
   *
   * The base implementation captures every output port value. Subclasses that
   * hold private loop counters or other transient state (e.g. ForLoopNode's
   * _loopIndex / _loopActive) should override this and spread super's result:
   *
   *   serializeState() {
   *     return { ...super.serializeState(), _loopIndex: this._loopIndex, _loopActive: this._loopActive };
   *   }
   */
  serializeState(): Record<string, unknown> {
    const portValues: Record<string, unknown> = {};
    for (const [k, p] of Object.entries(this.outputs)) {
      portValues[`out:${k}`] = (p as any).value;
    }
    for (const [k, p] of Object.entries(this.inputs)) {
      portValues[`in:${k}`] = (p as any).value;
    }
    return portValues;
  }

  /**
   * Restore internal execution state from a previously captured snapshot.
   * The base implementation restores port values. Subclasses should override
   * to restore their own private fields as well.
   */
  deserializeState(state: Record<string, unknown>): void {
    for (const [key, value] of Object.entries(state)) {
      if (key.startsWith('out:')) {
        const portName = key.slice(4);
        if (this.outputs[portName]) (this.outputs[portName] as any).setValue(value);
      } else if (key.startsWith('in:')) {
        const portName = key.slice(3);
        if (this.inputs[portName]) (this.inputs[portName] as any).setValue(value);
      }
    }
  }

  delete_input(portName: string): void {
    if (portName in this.inputs) {
      delete this.inputs[portName];
    }
  }

  delete_output(portName: string): void {
    if (portName in this.outputs) {
      delete this.outputs[portName];
    }
  }

  add_control_input(portName: string): InputControlPort {
    const port = new InputControlPort(this.id, portName);
    this.inputs[portName] = port;
    return port;
  }

  add_control_output(portName: string): OutputControlPort {
    const port = new OutputControlPort(this.id, portName);
    this.outputs[portName] = port;
    return port;
  }

  add_data_input(portName: string, dataType: ValueType = ValueType.ANY): InputDataPort {
    if (portName in this.inputs) {
      throw new Error(
        `Data input port '${portName}' already exists in node '${this.id}'`,
      );
    }
    const port = new InputDataPort(this.id, portName, dataType);
    this.inputs[portName] = port;
    return port;
  }

  add_data_output(portName: string, dataType: ValueType = ValueType.ANY): OutputDataPort {
    if (portName in this.outputs) {
      throw new Error(
        `Data output port '${portName}' already exists in node '${this.id}'`,
      );
    }
    const port = new OutputDataPort(this.id, portName, dataType);
    this.outputs[portName] = port;
    return port;
  }

  get_input_ports(restrictTo: PortFunction | null = null): NodePort[] {
    const all = Object.values(this.inputs);
    if (restrictTo === null) return all;
    return all.filter((p) => p.function === restrictTo);
  }

  get_output_ports(restrictTo: PortFunction | null = null): NodePort[] {
    const all = Object.values(this.outputs);
    if (restrictTo === null) return all;
    return all.filter((p) => p.function === restrictTo);
  }

  get_output_data_ports(): NodePort[] {
    return this.get_output_ports(PortFunction.DATA);
  }

  get_output_control_ports(): NodePort[] {
    return this.get_output_ports(PortFunction.CONTROL);
  }

  get_input_data_ports(): NodePort[] {
    return this.get_input_ports(PortFunction.DATA);
  }

  get_input_control_ports(): NodePort[] {
    return this.get_input_ports(PortFunction.CONTROL);
  }

  get_input_data_port(portName: string): NodePort {
    const port = this.inputs[portName];
    if (!port) throw new Error(`Input port '${portName}' not found in node '${this.id}'`);
    if (!port.isDataPort())
      throw new Error(
        `Input port '${portName}' in node '${this.id}' is not a data port`,
      );
    return port;
  }

  get_output_data_port(portName: string): NodePort {
    const port = this.outputs[portName];
    if (!port)
      throw new Error(`Output port '${portName}' not found in node '${this.id}'`);
    if (!port.isDataPort())
      throw new Error(
        `Output port '${portName}' in node '${this.id}' is not a data port`,
      );
    return port;
  }

  get_input_control_port(portName: string): NodePort {
    const port = this.inputs[portName];
    if (!port)
      throw new Error(`Input port '${portName}' not found in node '${this.id}'`);
    if (!port.isControlPort())
      throw new Error(
        `Input port '${portName}' in node '${this.id}' is not a control port`,
      );
    return port;
  }

  get_output_control_port(portName: string): NodePort {
    const port = this.outputs[portName];
    if (!port)
      throw new Error(`Output port '${portName}' not found in node '${this.id}'`);
    if (!port.isControlPort())
      throw new Error(
        `Output port '${portName}' in node '${this.id}' is not a control port`,
      );
    return port;
  }

  precompute(): void {
    console.log(`PRECOMPUTE: node '${this.id}'`);
  }

  postcompute(): void {
    this.dump_dirty_states();
    if (!this.all_data_inputs_clean()) {
      throw new Error(`postcompute: not all data inputs are clean on node '${this.id}'`);
    }
    if (!this.all_data_outputs_clean()) {
      throw new Error(`postcompute: not all data outputs are clean on node '${this.id}'`);
    }
    this._isDirty = false;
  }

  all_data_inputs_clean(): boolean {
    for (const port of this.get_input_data_ports()) {
      if (port._isDirty) return false;
    }
    return true;
  }

  all_data_outputs_clean(): boolean {
    for (const port of this.get_output_data_ports()) {
      if (port._isDirty) return false;
    }
    return true;
  }

  dump_dirty_states(): void {
    console.log(`  Node '${this.id}' port dirty states:`);
    const ctrlIns = this.get_input_control_ports();
    console.log(`.    Control Inputs: (${ctrlIns.length})`);
    for (const p of ctrlIns) {
      console.log(
        `.       Input Port [${(p as any).port_name}] dirty: ${p.isDirty()}; active: ${(p as any).isActive?.() ?? 'N/A'}`,
      );
    }
    const ctrlOuts = this.get_output_control_ports();
    console.log(`.    Control Outputs: (${ctrlOuts.length})`);
    for (const p of ctrlOuts) {
      console.log(
        `.       Output Port [${(p as any).port_name}] dirty: ${p.isDirty()}; active: ${(p as any).isActive?.() ?? 'N/A'}`,
      );
    }
    const dataIns = this.get_input_data_ports();
    console.log(`.    Data Inputs: (${dataIns.length})`);
    for (const p of dataIns) {
      console.log(`.       Input Port [${(p as any).port_name}] dirty: ${p.isDirty()}; value: ${(p as any).value}`);
    }
    const dataOuts = this.get_output_data_ports();
    console.log(`.    Data Outputs: (${dataOuts.length})`);
    for (const p of dataOuts) {
      console.log(`.       Output Port [${(p as any).port_name}] dirty: ${p.isDirty()}; value: ${(p as any).value}`);
    }
  }

  abstract compute(executionContext?: any): Promise<IExecutionResult>;

  compile(builder: any): void {
    // Placeholder
  }

  generate_IRC(): void {
    // Placeholder
  }
}
